Rubberduck.Parsing/VBA/ParseCoordinator.cs

```diff
@@ -58,105 +62,165 @@ public class ParseCoordinator : IParseCoordinator
+                }
 
-            if (!_isTestScope)
+                var originalStatus = State.Status;
+                if (!e.AllowedRunStates.Contains(originalStatus))
+                {
+                    e.Result = SuspensionResult.IncompatibleState;
+                    return;
+                }
+                _parserStateManager.SetStatusAndFireStateChanged(e.Requestor, ParserState.Busy,
+                    CancellationToken.None);
+                e.BusyAction.Invoke();
+            }
+            catch
             {
-                Task.Run(() => ParseAll(sender, token), token);
+                e.Result = SuspensionResult.UnexpectedError;
+                throw;
             }
-            else
+            finally
+            {
+                lock (SuspendStackSyncObject)
+                {
+                    _isSuspended = false;
+                    if (_requestorStack.TryPop(out var lastRequestor))
+                    {
+                        _requestorStack.Clear();
+                        parseRequestor = lastRequestor;
+                    }
+
+                    // Though there were no reparse requests, we need to reset the state before we release the 
+                    // write lock to avoid introducing discrepancy in the parser state due to readers being 
+                    // blocked. Any reparse requests must be done outside the write lock; see further below.
+                    if (parseRequestor == null)
+                    {
+                        // We cannot make any assumptions about the original state, nor do we know
+                        // anything about resuming the previous state, so we must delegate the state
+                        // evaluation to the state manager.
+                        _parserStateManager.EvaluateOverallParserState(CancellationToken.None);
+                    }
+                }
+
+                if (ParsingSuspendLock.IsWriteLockHeld)
+                {
+                    ParsingSuspendLock.ExitWriteLock();
+                }
+
+                if (e.Result == SuspensionResult.Pending)
+                {
+                    e.Result = SuspensionResult.Completed;
+                }
+            }
+
+            // Any reparse requests must be done outside the write lock to avoid deadlocks
+            if (parseRequestor != null)
             {
-                ParseInternal(token);
+                BeginParse(parseRequestor);
             }
         }
 
-        private void Cancel(bool createNewTokenSource = true)
+        /// <remarks>
+        /// Overriden in the unit test project to facilicate synchronous unit tests
+        /// Refer to TestParserCoordinator class in the unit test project.
+        /// </remarks>
+        protected virtual void BeginParse(object sender)
         {
-            lock (_cancellationSyncObject)
+            Task.Run(() => ParseAll(sender));
+        }
+
+        public void Cancel()
```

Comment:
Yes, I added that for the API because I realize it had no way to cancel a parse, which doesn't make sense since it can support running the parse async and receiving events for each parser state change. 

Based on your other comment, I agree that it need to be its own subclass.

For my learning, why this is bad? I thought since the coordinator still handles the cancellation internally, it should not have a problem. The matter of calling parse would be on the caller?

Reply:
The `ParseCoordinator` is set up to only ever finish in either the `Ready` state or the in an error state. This allows us to deactivate the refresh command in all other states, except `Pending` because it is the initial state.
Correspondingly, the working assumption is that a cancellation is only ever caused by a never parse request. In that case, it is safe to abandon the previous run in the middle because the next parse will run to the end, unless yet another parse gets requested. (It requires some thoughtful housekeeping to maintain the proper state of the coordinator to not forget to do stuff on the reparse.) The cancellation was never intended to be called from the outside. It is purely a performance enhancement when a reparse is requested while another parse is still running.